src/common/ast/ast.rs

```diff
@@ -0,0 +1,48 @@
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq)]
+pub struct Model {
+    pub variables: HashMap<Name, DecisionVariable>,
+    pub constraints: Vec<Expression>,
+}
+
+impl Model {
+    // Function to update a DecisionVariable based on its Name
+    pub fn update_domain(&mut self, name: &Name, new_domain: Domain) {
+        if let Some(decision_var) = self.variables.get_mut(name) {
+            decision_var.domain = new_domain;
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub enum Name {
+    UserName(String),
+    MachineName(i32),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DecisionVariable {
+    pub domain: Domain,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Domain {
+    BoolDomain,
+    IntDomain(Vec<Range<i32>>),
```

Comment:
This is fine for now, but at some point someone will have to write (or find) a lovely library for ranges of ranges -- at the moment the domains [0..3), [3..6), [0..3], [1..6) and [0..6) would all be different, despite them having the same underlying values. I'm sure at a later date there will be a demand for interesting + unioning such domains as well.

Reply:
fully agreed